Update BioGPT model card (#38214)
huggingface/transformers: docs/source/en/model_doc/biogpt.md

-->

<div style="float: right;">
    <div class="flex flex-wrap space-x-1">
            <img alt="PyTorch" src="https://img.shields.io/badge/PyTorch-DE3412?style=flat&logo=pytorch&logoColor=white">
            <img alt="SDPA" src="https://img.shields.io/badge/SDPA-DE3412?style=flat&logo=pytorch&logoColor=white">
    </div>
</div>

# BioGPT

> [BioGPT](https://huggingface.co/papers/2210.10341) is a generative Transformer model based on the GPT-2 architecture and pre-trained on 15 million PubMed abstracts. That makes it great at writing, understanding, and analyzing biomedical text.

Reviewer: ```suggestion
[BioGPT](https://huggingface.co/papers/2210.10341) is a generative Transformer model based on [GPT-2](./gpt2) and pretrained on 15 million PubMed abstracts. It is designed for biomedical language tasks.
```

> You can find all the original BioGPT checkpoints under the [BioGPT](https://huggingface.co/models?search=biogpt) collection.

Reviewer: ```suggestion
You can find all the original BioGPT checkpoints under the [Microsoft](https://huggingface.co/microsoft?search_models=biogpt) organization.
```

> [!TIP]

> > Click on the BioGPT models in the right sidebar for more examples of how to apply BioGPT to different biomedical text generation and question answering tasks.

Reviewer: ```suggestion
> Click on the BioGPT models in the right sidebar for more examples of how to apply BioGPT to different language tasks.
```

The example below demonstrates how to generate biomedical text with [`Pipeline`], [`AutoModel`], and also from the command line.

<hfoptions id="usage">
<hfoption id="Pipeline">

```py
import torch
from transformers import pipeline

generator = pipeline(
    task="text-generation",
    model="microsoft/biogpt",
    torch_dtype=torch.float16,

>     device_map="auto",  # 0 for GPU, -1 for CPU

Reviewer: ```suggestion
    device=0,
```

)
result = generator("Ibuprofen is best used for", truncation=True, max_length=50, do_sample=True)[0]["generated_text"]
print(result)
```

</hfoption>
<hfoption id="AutoModel">

```py

> import torch
> from transformers import AutoModelForCausalLM, AutoTokenizer
> 
> tokenizer = AutoTokenizer.from_pretrained("microsoft/biogpt")
> model = AutoModelForCausalLM.from_pretrained(
>     "microsoft/biogpt",
>     torch_dtype=torch.float16,
>     device_map="auto", # 0 for GPU, -1 for CPU
>     attn_implementation="sdpa"
> )
> input_text = "Ibuprofen is best used for"
> inputs = tokenizer(input_text, return_tensors="pt").to(model.device)
> with torch.no_grad():
>     generated_ids = model.generate(**inputs, max_length=50)
> output = tokenizer.decode(generated_ids[0], skip_special_tokens=True)
> print(output)

Reviewer: ```suggestion
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("microsoft/biogpt")
model = AutoModelForCausalLM.from_pretrained(
    "microsoft/biogpt",
    torch_dtype=torch.float16,
    device_map="auto",
    attn_implementation="sdpa"
)

input_text = "Ibuprofen is best used for"
inputs = tokenizer(input_text, return_tensors="pt").to("cuda)

with torch.no_grad():
    generated_ids = model.generate(**inputs, max_length=50)
    
output = tokenizer.decode(generated_ids[0], skip_special_tokens=True)
print(output)
```

Author: I noticed a little hiccup, by using `device_map="auto"` and then `inputs = tokenizer(input_text, return_tensors="pt").to("cuda)`, if the code is running on the CPU, the `.to("cuda") would make an error.

```

</hfoption>

> <hfoption id="transformers-cli">

Reviewer: ```suggestion
<hfoption id="transformers CLI">
```

```bash

> echo -e "Ibuprofen is best used for" \
> | transformers run --task text-generation \
>                    --model microsoft/biogpt \
>                    --device -1

Reviewer: ```suggestion
echo -e "Ibuprofen is best used for" | transformers run --task text-generation --model microsoft/biogpt --device 0
```

Reviewer: Sorry, this should actually still be `transformers-cli` for the `run` command!

```

</hfoption>
</hfoptions>

> 

Reviewer: Let's also include a quantization example using https://huggingface.co/microsoft/BioGPT-Large

Author: I'm a total noob on quantization but from what I researched, is this something like this :

```py
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16,
    bnb_4bit_use_double_quant=True
)

tokenizer = AutoTokenizer.from_pretrained("microsoft/BioGPT-Large")
model = AutoModelForCausalLM.from_pretrained(
    "microsoft/BioGPT-Large", 
    quantization_config=bnb_config,
    torch_dtype=torch.bfloat16,
    device_map="auto"
)

input_text = "Ibuprofen is best used for"
inputs = tokenizer(input_text, return_tensors="pt").to(model.device)
with torch.no_grad():
    generated_ids = model.generate(**inputs, max_length=50)    
output = tokenizer.decode(generated_ids[0], skip_special_tokens=True)
print(output)
```

## Notes

> - BioGPT is a model with absolute position embeddings so it's usually advised to pad the inputs on the right rather than the left.

Reviewer: ```suggestion
- Pad inputs on the right because BioGPT uses absolute position embeddings.
```

> - BioGPT was trained with a causal language modeling (CLM) objective and is therefore powerful at predicting the next token in a sequence. Leveraging this feature allows BioGPT to generate syntactically coherent text as it can be observed in the run_generation.py example script.

Reviewer: ```suggestion
```

> - The model can take the `past_key_values` (for PyTorch) as input, which is the previously computed key/value attention pairs. Using this (past_key_values or past) value prevents the model from re-computing pre-computed values in the context of text generation. For PyTorch, see past_key_values argument of the BioGptForCausalLM.forward() method for more information on its usage.

Reviewer: ```suggestion
- BioGPT can reuse previously computed key-value attention pairs. Access this feature with the [past_key_values](https://huggingface.co/docs/transformers/main/en/model_doc/biogpt#transformers.BioGptModel.forward.past_key_values) parameter in [`BioGPTModel.forward`].
```

> - The `head_mask` argument is ignored when using all attention implementation other than "eager". If you have a `head_mask` and want it to have effect, load the model with `XXXModel.from_pretrained(model_id, attn_implementation="eager")`

Reviewer: ```suggestion
- The `head_mask` argument is ignored when using an attention implementation other than "eager". If you want to use `head_mask`, make sure `attn_implementation="eager"`).

   ```py
   from transformers import AutoModelForCausalLM
   
   model = AutoModelForCausalLM.from_pretrained(
      "microsoft/biogpt",
      attn_implementation="eager"
   )
```
